titan-compiler/coder.lua

```diff
@@ -197,19 +345,151 @@ generate_program = function(prog, modname)
         MODNAME = modname,
+        N_TOPLEVEL = c_integer(n_toplevel),
+        DEFINE_FUNCTIONS = define_functions,
         INITIALIZE_TOPLEVEL = initialize_toplevel,
         CREATE_MODULE_TABLE = create_module_table,
     })
     return pretty.reindent_c(code)
 end
 
--- Returns (statements, value)
+
+generate_stat = function(stat)
+    local tag = stat._tag
+    if     tag == ast.Stat.Block then
+        local cstatss = {}
+        table.insert(cstatss, "{")
+        for _, inner_stat in ipairs(stat.stats) do
+            local cstats = generate_stat(inner_stat)
+            table.insert(cstatss, cstats)
+        end
+        table.insert(cstatss, "}")
+        return table.concat(cstatss, "\n")
+
+    elseif tag == ast.Stat.While then
+        local cond_cstats, cond_cvalue = generate_exp(stat.condition)
+        local block_cstats = generate_stat(stat.block)
+        if cond_cstats == "" then
+            return util.render([[
+                while(${COND}) ${BLOCK}
+            ]], {
+                COND = cond_cvalue,
+                BLOCK = block_cstats
+            })
+        else
+            return util.render([[
+                for(;;) {
+                    ${COND_STATS}
+                    if (!(${COND})) break;
+                    ${BLOCK}
+                }
+            ]], {
+                COND_STATS = cond_cstats,
+                COND = cond_cvalue,
+                CLOCK = block_cstats
+            })
+        end
+
+    elseif tag == ast.Stat.Repeat then
+        error("not implemented yet")
+
+    elseif tag == ast.Stat.If then
+        error("not implemented yet")
+
+    elseif tag == ast.Stat.For then
+        error("not implemented yet")
+
+    elseif tag == ast.Stat.Assign then
+        local var_cstats, var_clvalue = generate_var(stat.var)
+        local exp_cstats, exp_cvalue = generate_exp(stat.exp)
+        return util.render([[
+            ${VAR_STATS}
+            ${EXP_STATS}
+            ${LVALUE} = ${RVALUE};
+        ]], {
+            VAR_STATS = var_cstats,
+            EXP_STATS = exp_cstats,
+            LVALUE = var_clvalue,
+            RVALUE = exp_cvalue
+        })
+
+
+    elseif tag == ast.Stat.Decl then
+        local exp_cstats, exp_cvalue = generate_exp(stat.exp)
+        return util.render([[
+            ${STATS}
+            ${DECLARATION} = ${VALUE};
+        ]], {
+            STATS = exp_cstats,
+            VALUE = exp_cvalue,
+            DECLARATION = c_declaration(ctype(stat.decl._type), local_name(stat.decl.name))
+        })
+
+    elseif tag == ast.Stat.Call then
+        error("not implemented yet")
+
+    elseif tag == ast.Stat.Return then
+        local cstats, cvalue = generate_exp(stat.exp)
+        return util.render([[
+            ${CSTATS}
+            return ${CVALUE};
+        ]], {
+            CSTATS = cstats,
+            CVALUE = cvalue
+        })
+
+    else
+        error("impossible")
+    end
+end
+
+-- @returns (statements, clvalue)
```

Comment:
Is `clvalue` a C lvalue? Could you rename the comment to c_lvalue if this is the case?